src/hypofuzz/provider.py

```diff
@@ -26,29 +80,409 @@ def fresh_choice(
+
+            choices = mutator.generate_choices()
+            self._start_phase(Phase.GENERATE)
+
+        start = time.perf_counter()
+        self.ninputs += 1
+
+        self._state = State(
+            choices=choices,
+            replay_priority=replay_priority,
+            start_time=start,
+            extra_queue_data=extra_queue_data,
+        )
+
+    def on_observation(
+        self, observation: Union[TestCaseObservation, InfoObservation]
+    ) -> None:
+        assert observation.type == "test_case"
+        self.after_test_case(observation)
+
+    def after_test_case(self, observation: TestCaseObservation) -> None:
+        assert self._state is not None
+        assert self._state.branches is not None
+        assert self.corpus is not None
+        assert self.database_key is not None
+        assert self.db is not None
+
+        assert observation.type == "test_case"
+        # run_start is relative to StateForActualGivenExecution, which we
+        # re-use per FuzzProcess. Overwrite with the current timestamp for use
+        # in sorting observations. This is not perfectly reliable in a
+        # distributed setting, but is good enough.
+        observation.run_start = time.time()
+        # "arguments" duplicates part of the call repr in "representation".
+        # We don't use this for anything, and it can be substantial in size, so
+        # drop it.
+        observation.arguments = {}
+
+        status = observation.metadata.data_status
+        self.elapsed_time += time.perf_counter() - self._state.start_time
+        self.status_counts[status] += 1
+
+        if status is not Status.INTERESTING and self._state.replay_priority in [
+            ReplayPriority.FAILURE_SHRUNK,
+            ReplayPriority.FAILURE_UNSHRUNK,
+        ]:
+            # this failure didn't reproduce. It's either been fixed, is flaky,
+            # or is specific to the worker's environment (e.g. only fails on
+            # python 3.11, while this worker is on python 3.12).
+            # In any case, remove it from the db.
+            #
+            # We know whether this failure was shrunk or not as of when we
+            # took it out of the db (via replay_priority). But I'm not confident that
+            # it's not possible for another worker to move the same choice
+            # sequence from unshrunk to shrunk - and failures are rare +
+            # deletions cheap. Just try deleting from both.
+            self.db.delete_failure(self.database_key, self._state.choices, shrunk=True)
+            self.db.delete_failure(self.database_key, self._state.choices, shrunk=False)
+            failure_observation = self._state.extra_queue_data
+            assert failure_observation is not None
+            assert isinstance(failure_observation, Observation)
+            self.db.delete_failure_observation(
+                self.database_key, self._state.choices, failure_observation
+            )
```

Comment:
followup for a future PR: let's evict shrunk failures only if the observation timestamp is 8+ days ago